File: tron-backend/src/adapters/github.js
const axios = require('axios');

const GITHUB_TOKEN = process.env.GITHUB_ACCESS_TOKEN;

const githubAPI = axios.create({
    headers: {
        'Authorization': `Bearer ${GITHUB_TOKEN}`,
        'Accept': 'application/vnd.github.v3.diff', // This tells GitHub we want raw diff text, not JSON!
        'User-Agent': 'TRON-AI-Pipeline'
    }
});

// The "Monster Diff" Sanitizer (REQ-9 & REQ-10)
async function fetchAndSanitizeDiff(diffUrl) {
    console.log(`\n📥 [GITHUB ADAPTER] Fetching raw code diff from: ${diffUrl}`);

    if (!GITHUB_TOKEN) {
        throw new Error("Missing GITHUB_ACCESS_TOKEN in .env");
    }

    try {
        const response = await githubAPI.get(diffUrl);
        let rawDiff = response.data;

        console.log(`🧹 [GITHUB ADAPTER] Sanitizing Monster Diffs...`);

        // 🛡️ SECURITY FIX: The Secret Scrubber (DLP)
        const secretPatterns = [
            /(A3T[A-Z0-9]|AKIA|AGPA|AIDA|AROA|AIPA|ANPA|ANVA|ASIA)[A-Z0-9]{16}/g, // AWS Access Keys
            /-----BEGIN (RSA|EC|DSA|OPENSSH) PRIVATE KEY-----\n[\s\S]*?\n-----END \1 PRIVATE KEY-----/g, // SSH/RSA Private Keys
            /(password|secret|api_key|access_token|client_secret)[ \t]*[:=][ \t]*['"]?[^\s'"]+['"]?/gi // Generic hardcoded secrets
        ];

        let originalLength = rawDiff.length;
        secretPatterns.forEach(pattern => {
            rawDiff = rawDiff.replace(pattern, '[REDACTED_SECRET]');
        });

        if (rawDiff.length !== originalLength) {
            console.warn(`🚨 [GITHUB ADAPTER] WARNING: Secrets detected and redacted from PR diff!`);
        }

        // 1. Remove lockfiles (package-lock.json, yarn.lock)
        rawDiff = rawDiff.replace(/diff --git a\/.*(?:package-lock\.json|yarn\.lock)[\s\S]*?(?=diff --git|$)/g, '');
        
        // 2. Remove minified files and SVGs
        rawDiff = rawDiff.replace(/diff --git a\/.*\.(?:min\.js|svg|map)[\s\S]*?(?=diff --git|$)/g, '');

        // 3. Enforce Hard Token Limits (Approx. 12,000 characters to stay safe for LLMs)
        const MAX_CHARS = 12000;
        if (rawDiff.length > MAX_CHARS) {
            console.warn(`⚠️ [GITHUB ADAPTER] Diff exceeds ${MAX_CHARS} characters. Truncating to protect LLM limit.`);
            rawDiff = rawDiff.substring(0, MAX_CHARS) + '\n\n... [DIFF TRUNCATED BY T.R.O.N. DUE TO LENGTH] ...';
        }

        console.log(`✅ [GITHUB ADAPTER] Diff sanitized. Final length: ${rawDiff.length} characters.`);
        return rawDiff;

    } catch (error) {
        console.error(`❌ [GITHUB ADAPTER] Failed to fetch diff from GitHub.`);
        throw error;
    }
}

// 📢 NEW: Post a comment directly to a Pull Request
async function postPullRequestComment(repoFullName, prNumber, commentBody) {
    if (!GITHUB_TOKEN) {
        console.warn("⚠️ GITHUB_ACCESS_TOKEN is missing in .env. Cannot post PR comment.");
        return;
    }

    // GitHub's API endpoint for PR comments is actually the Issues endpoint!
    const url = `https://api.github.com/repos/${repoFullName}/issues/${prNumber}/comments`;
    
    const headers = {
        'Authorization': `Bearer ${GITHUB_TOKEN}`,
        'Accept': 'application/vnd.github.v3+json', // We need JSON for this call, not diff
        'Content-Type': 'application/json',
        'User-Agent': 'TRON-AI-Pipeline'
    };

    try {
        await axios.post(url, { body: commentBody }, { headers });
        console.log(`💬 Successfully posted AI review to PR #${prNumber}`);
    } catch (error) {
        console.error(`❌ [GITHUB ADAPTER] Failed to post PR comment: ${error.message}`);
        if (error.response) {
            console.error(`   -> Details:`, JSON.stringify(error.response.data, null, 2));
        }
    }
}

// Ensure both functions are exported!
module.exports = { fetchAndSanitizeDiff, postPullRequestComment };

File: tron-go-backend/internal/adapters/github.go
package adapters

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"regexp"
	"strings"
	"time"

	"github.com/tron-v3.1/tron-go-backend/internal/services"
)

type GitHubAdapter struct {
	HTTPClient *http.Client
}

func NewGitHubAdapter() *GitHubAdapter {
	return &GitHubAdapter{
		HTTPClient: &http.Client{Timeout: 15 * time.Second},
	}
}

// FetchAndSanitizeDiff fetches the raw code and runs the DLP / Secret Scrubber (REQ-9 & 10)
func (api *GitHubAdapter) FetchAndSanitizeDiff(repoFullName string, prNumber int, installationID string) (string, error) {
	fmt.Printf("\n📥 [GITHUB ADAPTER] Fetching raw code diff for PR #%d\n", prNumber)

	token, err := services.GetInstallationToken(installationID)
	if err != nil {
		return "", fmt.Errorf("failed to get github token: %w", err)
	}

	url := fmt.Sprintf("https://api.github.com/repos/%s/pulls/%d", repoFullName, prNumber)
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return "", err
	}

	req.Header.Set("Authorization", "Bearer "+token)
	req.Header.Set("Accept", "application/vnd.github.v3.diff")
	req.Header.Set("User-Agent", "TRON-AI-Pipeline")

	resp, err := api.HTTPClient.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	diffBytes, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}

	rawDiff := string(diffBytes)
	fmt.Println("🧹 [GITHUB ADAPTER] Sanitizing Monster Diffs...")

	// 1. Remove Lockfiles, SVGs, and Minified files safely by splitting the diff per file
	fileChunks := strings.Split(rawDiff, "diff --git")
	var cleanDiffBuilder strings.Builder

	for _, chunk := range fileChunks {
		if chunk == "" {
			continue
		}

		// If it's a lockfile or minified file, drop the entire chunk
		if strings.Contains(chunk, "package-lock.json") ||
			strings.Contains(chunk, "yarn.lock") ||
			strings.Contains(chunk, ".min.js") ||
			strings.Contains(chunk, ".svg") ||
			strings.Contains(chunk, ".map") {
			continue
		}

		cleanDiffBuilder.WriteString("diff --git")
		cleanDiffBuilder.WriteString(chunk)
	}

	sanitizedDiff := cleanDiffBuilder.String()
	originalLength := len(sanitizedDiff)

	// 2. The Secret Scrubber (DLP)
	awsRe := regexp.MustCompile(`(A3T[A-Z0-9]|AKIA|AGPA|AIDA|AROA|AIPA|ANPA|ANVA|ASIA)[A-Z0-9]{16}`)
	sshRe := regexp.MustCompile(`(?s)-----BEGIN (RSA|EC|DSA|OPENSSH) PRIVATE KEY-----.*?-----END \1 PRIVATE KEY-----`)
	genericRe := regexp.MustCompile(`(?i)(password|secret|api_key|access_token|client_secret)[ \t]*[:=][ \t]*['"]?[^\s'"]+['"]?`)

	sanitizedDiff = awsRe.ReplaceAllString(sanitizedDiff, "[REDACTED_SECRET]")
	sanitizedDiff = sshRe.ReplaceAllString(sanitizedDiff, "[REDACTED_SECRET]")
	sanitizedDiff = genericRe.ReplaceAllString(sanitizedDiff, "[REDACTED_SECRET]")

	if len(sanitizedDiff) != originalLength {
		fmt.Println("🚨 [GITHUB ADAPTER] WARNING: Secrets detected and redacted from PR diff!")
	}

	// 3. Enforce Hard Token Limits (12,000 chars)
	const maxChars = 12000
	if len(sanitizedDiff) > maxChars {
		fmt.Printf("⚠️ [GITHUB ADAPTER] Diff exceeds %d characters. Truncating to protect LLM limit.\n", maxChars)
		sanitizedDiff = sanitizedDiff[:maxChars] + "\n\n... [DIFF TRUNCATED BY T.R.O.N. DUE TO LENGTH] ..."
	}

	fmt.Printf("✅ [GITHUB ADAPTER] Diff sanitized. Final length: %d characters.\n", len(sanitizedDiff))
	return sanitizedDiff, nil
}

// PostPullRequestComment adds the AI review to the PR
func (api *GitHubAdapter) PostPullRequestComment(repoFullName string, prNumber int, commentBody, installationID string) error {
	token, err := services.GetInstallationToken(installationID)
	if err != nil {
		return err
	}

	url := fmt.Sprintf("https://api.github.com/repos/%s/issues/%d/comments", repoFullName, prNumber)
	payload := map[string]string{"body": commentBody}
	payloadBytes, _ := json.Marshal(payload)

	req, err := http.NewRequest("POST", url, bytes.NewBuffer(payloadBytes))
	if err != nil {
		return err
	}

	req.Header.Set("Authorization", "Bearer "+token)
	req.Header.Set("Accept", "application/vnd.github.v3+json")
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("User-Agent", "TRON-AI-Pipeline")

	resp, err := api.HTTPClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusCreated {
		return fmt.Errorf("failed to post comment, status: %d", resp.StatusCode)
	}

	fmt.Printf("💬 Successfully posted AI review to PR #%d\n", prNumber)
	return nil
}
